pkg/reconciler/timeout_handler.go

```diff
@@ -0,0 +1,219 @@
+	var finished chan bool
+	key := runObj.GetRunKey()
+	t.doneMut.Lock()
+	if existingfinishedChan, ok := t.done[key]; ok {
+		finished = existingfinishedChan
+	} else {
+		finished = make(chan bool)
+	}
+	t.done[key] = finished
+	t.doneMut.Unlock()
+	return finished
+}
+
+func (t *TimeoutSet) StatusLock(runObj StatusKey) {
+	m, _ := t.statusMap.LoadOrStore(runObj.GetRunKey(), &sync.Mutex{})
+	mut := m.(*sync.Mutex)
+	mut.Lock()
+}
+
+func (t *TimeoutSet) StatusUnlock(runObj StatusKey) {
+	m, ok := t.statusMap.Load(runObj.GetRunKey())
+	if !ok {
+		return
+	}
+	mut := m.(*sync.Mutex)
+	mut.Unlock()
+}
+
+func getTimeout(d *metav1.Duration) time.Duration {
+	timeout := defaultTimeout
+	if d != nil {
+		timeout = d.Duration
+	}
+	return timeout
+}
+
+// checkPipelineRunTimeouts function creates goroutines to wait for pipelinerun to
+// finish/timeout in a given namespace
+func (t *TimeoutSet) checkPipelineRunTimeouts(namespace string) {
+	pipelineRuns, err := t.pipelineclientset.TektonV1alpha1().PipelineRuns(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		t.logger.Errorf("Can't get taskruns list in namespace %s: %s", namespace, err)
+	}
+	for _, pipelineRun := range pipelineRuns.Items {
+		pipelineRun := pipelineRun
+		if pipelineRun.IsDone() || pipelineRun.IsCancelled() {
+			continue
+		}
+		go t.WaitPipelineRun(&pipelineRun)
+	}
+}
+
+// CheckTimeouts function iterates through all namespaces and calls corresponding
+// taskrun/pipelinerun timeout functions
+func (t *TimeoutSet) CheckTimeouts() {
+	namespaces, err := t.kubeclientset.CoreV1().Namespaces().List(metav1.ListOptions{})
+	if err != nil {
+		t.logger.Errorf("Can't get namespaces list: %s", err)
+	}
+	for _, namespace := range namespaces.Items {
+		t.checkTaskRunTimeouts(namespace.GetName())
+		t.checkPipelineRunTimeouts(namespace.GetName())
+	}
+}
+
+// checkTaskRunTimeouts function creates goroutines to wait for pipelinerun to
+// finish/timeout in a given namespace
+func (t *TimeoutSet) checkTaskRunTimeouts(namespace string) {
+	taskruns, err := t.pipelineclientset.TektonV1alpha1().TaskRuns(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		t.logger.Errorf("Can't get taskruns list in namespace %s: %s", namespace, err)
+	}
+	for _, taskrun := range taskruns.Items {
+		taskrun := taskrun
+		if taskrun.IsDone() || taskrun.IsCancelled() {
+			continue
+		}
+		go t.WaitTaskRun(&taskrun)
+	}
+}
+
+// WaitTaskRun function creates a blocking function for taskrun to wait for
+// 1. Stop signal, 2. TaskRun to complete or 3. Taskrun to time out
+func (t *TimeoutSet) WaitTaskRun(tr *v1alpha1.TaskRun) {
+	timeout := getTimeout(tr.Spec.Timeout)
+	runtime := time.Duration(0)
+
+	t.StatusLock(tr)
+	if tr.Status.StartTime != nil && !tr.Status.StartTime.Time.IsZero() {
+		runtime = time.Since(tr.Status.StartTime.Time)
+	}
+	t.StatusUnlock(tr)
+	timeout -= runtime
+	finished := t.getOrCreateFinishedChan(tr)
+
+	defer t.Release(tr)
+
+	select {
+	case <-t.stopCh:
+	case <-finished:
```

Comment:
readability nit: empty case statements can make it hard to distinguish between "match any of these cases" and "match this case and do nothing"

I find it helpful to add a comment describing no-op cases, and/or explicitly return if you can, e.g.:

```
select {
  case <-t.stopCh:
    // we're stopping, give up
    return
  case <-finished:
    // it finished, we can stop watching
    return
  case <-time.After(timeout):
    ...
}
```